### PR TITLE
Add functionality to change Net::SSH's client identification string

### PR DESCRIPTION
**Pull request by wvu**
We'll want to send a patch to upstream later to make this easier.

## `Msf::Exploit::Remote::SSH`

```
msf5 > use libssh

Matching Modules
================

   #  Name                                      Disclosure Date  Rank    Check  Description
   -  ----                                      ---------------  ----    -----  -----------
   0  auxiliary/scanner/ssh/libssh_auth_bypass  2018-10-16       normal  Yes    libssh Authentication Bypass Scanner


[*] Using auxiliary/scanner/ssh/libssh_auth_bypass
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > setg rhosts 127.0.0.1
rhosts => 127.0.0.1
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > setg rport 2222
rport => 2222
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > grep SSH advanced
   SSH_DEBUG                   false            no        SSH debugging
   SSH_IDENT                   test             yes       SSH client identification string
   SSH_TIMEOUT                 10               no        SSH timeout
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > run

[*] 127.0.0.1:2222 - Attempting authentication bypass
^C[*] Caught interrupt from the console...
[*] Auxiliary module execution completed
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) >
```

```
wvu@kharak:~$ ncat -lkv 2222
Ncat: Version 7.70 ( https://nmap.org/ncat )
Ncat: Listening on :::2222
Ncat: Listening on 0.0.0.0:2222
Ncat: Connection from 127.0.0.1.
Ncat: Connection from 127.0.0.1:58371.
SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.3
```

```
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > setg ssh_ident SSH-2.0-OpenSSH_7.9
ssh_ident => SSH-2.0-OpenSSH_7.9
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > run

[*] 127.0.0.1:2222 - Attempting authentication bypass
```

```
Ncat: Connection from 127.0.0.1.
Ncat: Connection from 127.0.0.1:58381.
SSH-2.0-OpenSSH_7.9
```

## `Metasploit::Framework::LoginScanner::SSH`

```
msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > use ssh_login

Matching Modules
================

   #  Name                                    Disclosure Date  Rank    Check  Description
   -  ----                                    ---------------  ----    -----  -----------
   0  auxiliary/scanner/ssh/ssh_login                          normal  Yes    SSH Login Check Scanner
   1  auxiliary/scanner/ssh/ssh_login_pubkey                   normal  Yes    SSH Public Key Login Scanner


msf5 auxiliary(scanner/ssh/libssh_auth_bypass) > use 0
msf5 auxiliary(scanner/ssh/ssh_login) > set username mulder
username => mulder
msf5 auxiliary(scanner/ssh/ssh_login) > set password trustno1
password => trustno1
msf5 auxiliary(scanner/ssh/ssh_login) > run
```

```
Ncat: Connection from 127.0.0.1.
Ncat: Connection from 127.0.0.1:58392.
SSH-2.0-OpenSSH_7.9
```

### CODE REVIEW

File: lib/msf/ui/console/driver.rb [hunk 1]
```diff
@@ -568,6 +570,33 @@ def handle_loglevel(val)
...
+  #
+  def handle_sshversion(val)
+    return false unless val && val.kind_of?(String) && !val.empty?
+
+    begin
+      require 'net/ssh'
+    rescue LoadError
+      return false
+    end
+
```
Comment (busterb) [inline lib/msf/ui/console/driver.rb]: The only way you'll be able to get rid of the hacks is with a PR to https://github.com/net-ssh/net-ssh. Looks live enough!

**Review by sempervictus — COMMENTED** [review]
That's why I still keep a net/ssh in lib. The original code was structured too much for client ops, not as a protocol + client setup. The intent is great though.

**Comment by wvu-r7** [comment]
I think we'll create our own R7 gem and hook it in.

**Comment by wvu-r7** [comment]
We've decided to create the gem, make changes there, and upstream them. Thanks!

**Comment by wvu-r7** [comment]
Bug.

**Review by busterb — APPROVED** [review]

**Comment by busterb** [comment]
I did notice that ssh_login is terribly slow to exit even if you just have one user/password, but that was unrelated to this PR. On the wire, the behavior is better as well because before it wouldn't 'FIN', but would just hang open. But that's stuff that could be improved later. Thanks!

**Comment by busterb** [comment]
## Release Notes

This adds additional options to the Metasploit SSH client libraries to simulate the wire signature of other clients.

**Comment by wvu-r7** [comment]
I noticed that, too. I believe `ssh_login`'s `SSH_TIMEOUT` is actually 30 seconds. I set the default to 10, though I allow modules to set their own for backward compatibility. Something to consider in our next refactor.

**Pull request merged**